### ISSUE TITLE
Drop PHPCS 2.x support by August 31, 2018

### ISSUE THREAD

**Issue by jrfnl** [WordPress-Coding-Standards/WordPress-Coding-Standards#1048]
As proposed in this comment https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/718#issuecomment-316300210, PHPCS 2.x should be dropped by August 31, 2018 at the latest.

That gives other external standards which build onto WPCS a year's time to upgrade their standards.

This issue is intended to document what needs to be done to drop PHPCS 2.x support, based on the PHPCS cross-version implementation as pulled in #1047.

Prerequisite: The minimum supported PHPCS version for WPCS will need to be `3.1.0` or higher.

### Sniff files:
- [x] Change namespace from `namespace WordPress...` to `namespace WordPressCS\WordPress`.
- [x] Change the `use` statements which refer to PHPCS classes to use the PHPCS 3.x names:
    ```php
    // Current:
    use PHP_CodeSniffer_Sniff as PHPCS_Sniff;
    use PHP_CodeSniffer_File as File;
    use PHP_CodeSniffer_Tokens as Tokens;

    // Should become:
    use PHP_CodeSniffer\Sniffs\Sniff as PHPCS_Sniff;
    use PHP_CodeSniffer\Files\File;
    use PHP_CodeSniffer\Util\Tokens;
    ```

### Rulesets:
- [x] All: remove the `<autoload>./PHPCSAliases.php</autoload>` statement.
- [x] `./WordPress/ruleset.xml`: adjust `<ruleset name="WordPress" namespace="WordPress">` to `<ruleset name="WordPress" namespace="WordPressCS\WordPress">`
- [x] Review which customizations currently done per error code can now be done for the complete sniff in one go. Related #1144 
- [x] Review other notes in the ruleset about additional customizations / sniff replacements which can be done once the minimum requirement has gone up & action them

### Cleanup:
- [x] Remove the `./WordPress/PHPCSAliases.php` file.
- [ ] ~~Remove the `./phpunit.xml.dist` file.~~
- [x] Remove the `./Test` subdirectory and all the files therein. The bootstraps and the PHPCS 2.x test suite files are no longer needed as long as the minimum PHPCS version goes up to 3.1.0.
- [x] Decide whether to keep or remove the `./WordPress/PHPCSHelper.php` file. It might be useful to leave it in place to be used for future PHPCS cross-version compatibility issues. The PHPCS 2.x specific code in the file can be removed.

### Build script and config files:
- [x] `.travis.yml`: Remove the PHPCS 2.x specific PHPUnit commands.
- [x] `.travis.yml`: Stop testing against PHP 5.3 (not supported by PHPCS 3.x).
- [x] `.travis.yml`: Stop testing against PHPCS 2.x branches.
- [x] `.travis.yml`: Remove the toggle for the `PHPCS_BIN` location.
- [x] `.travis.yml`: Limit the downloading of a custom PHPUnit version to HHVM, PHPCS supports PHPUnit 6.x since PHPCS 3.1.0. See #1130
- [x] `.travis,yml`: If needed: add the PHPCS native bootstrap to the PHPUnit command and/or the `phpunit.xml.dist` file. See https://github.com/squizlabs/PHP_CodeSniffer/pull/1384
- [x] Update the minimum required PHPCS version in the `composer.json` file.

### Documentation:
- [x] Readme: update minimum requirements & remove references to PHPCS 2.x.
- [x] Contributing: update the unit test instructions.
- [x] Remove the [PHPCS 3 and testing individual sniffs](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/PHPCS-3-and-testing-individual-sniffs) wiki page as it should no longer apply.

### Dropping PHP 5.3 support:
- [x] Remove code related to the `short_open_tag` ini setting from the `XSS.EscapeOutput` sniff.
- [x] Remove the short_open_tag related line from `.travis.yml`.
- [x] Remove unit tests work-around for `short_open_tag` ini setting icw PHP 5.3 in the `CodeAnalysis.EmptyStatement`, `Files.FileName`, `WhiteSpace.PrecisionAlignment` sniffs. See #1273, #1274

**Comment by jrfnl** [comment]
As a side-note: any WPCS sniff which would be pulled upstream over the next year can not (yet) be dropped from WPCS as the upstream version will only go into PHPCS 3.x.

So, if any WPCS sniffs would be upstreamed:
- [ ] Deprecate / Remove any sniffs of which there is now an upstream version and replace ruleset references to the WPCS version with the upstream sniff name.

**Comment by jrfnl** [comment]
Another one where a WPCS work-around may be removed if PHPCS merges the upstream fix: https://github.com/squizlabs/PHP_CodeSniffer/pull/1689 / #1185

**Comment by jrfnl** [comment]
A readme change which can be made once the minimum version has gone up to PHPCS `3.1.0` or higher:
* Recommend naming a custom project ruleset `.phpcs.xml.dist` (note the `.` at the start of the filename)

**Comment by jrfnl** [comment]
The PHPCS native ignore annotations will be changed in v3.2.0. The old type of annotations will be deprecated and they will be removed in PHPCS 4.x.

If WPCS would raise it's minimum PHPCS requirement to `3.2.0` when the work to drop PHPCS 2.x support will be done, this would be a good moment to address the change in annotations in all (unit test case) files.

It would also be a good moment to evaluate whether and if so, when to drop/deprecate the WPCS native whitelist comments.

From the upstream changelog:
```
  - This release deprecates the @codingStandards comment syntax used for sending commands to PHP_CodeSniffer
    -- The existing sytax will continue to work in all version 3 releases, but will be removed in version 4
    -- The comment formats have been replaced by a shorter syntax:
       --- @codingStandardsIgnoreFile becomes phpcs:ignoreFile
       --- @codingStandardsIgnoreStart becomes phpcs:disable
       --- @codingStandardsIgnoreEnd becomes phpcs:enable
       --- @codingStandardsIgnoreLine becomes phpcs:ignore
       --- @codingStandardsChangeSetting becomes phpcs:set
```

For more details:
* https://github.com/squizlabs/PHP_CodeSniffer/issues/604
* https://github.com/squizlabs/PHP_CodeSniffer/commit/532517044940717548518f887b76d42cca7486d6
* https://github.com/squizlabs/PHP_CodeSniffer/commit/389e52f3be59b28da1f592437e3ee254140dfc28
* https://github.com/squizlabs/PHP_CodeSniffer/commit/1f07a69c00ef3b933f295f6c047e297b679677e7

**Comment by jrfnl** [comment]
The `NamingConventions.ValidVariableName` sniff currently contains a `$php_reserved_vars` property. Once upstream PR squizlabs/PHP_CodeSniffer#1888 has been merged, that property can be removed and the upstream `$phpReservedVars` property can be used instead.

**Comment by jrfnl** [comment]
PHPCS upstream introduces a new `Generic.WhiteSpace.LanguageConstructSpacing` sniff to replace the `Squiz.WhiteSpace.LanguageConstructSpacing` sniff in v 3.3.0.
Once the minimum requirement of WPCS has gone up to `3.3.0`, we should switch over.

Also see: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/1328

**Comment by jrfnl** [comment]
More upstream goodies:

A PR has just been merged and will be included in the 3.3.0 release which changes the array format for properties.
The old format will remain valid as well for now, but is expected to be removed in v4.

If/when the minimum PHPCS version WPCS support goes up to 3.3.0, we should:
- [ ] Adjust the `array` properties used in the WPCS ruleset (if any).
- [ ] Adjust the wiki page regarding the WPCS custom properties and update all example code for array properties, such as the `text-domain`.
- [ ] Advise people through the changelog that the new array format is recommended for their own custom rulesets.

Old format:
```xml
<property name="forbiddenFunctions" type="array" value="print=>echo,create_function=>null" />
```

New format:
```xml
<property name="forbiddenFunctions" type="array">
    <element key="print" value="echo"/>
    <element key="create_function" value="null"/>
</property>
```

See: https://github.com/squizlabs/PHP_CodeSniffer/pull/1665

**Comment by GaryJones** [comment]
When we drop 2.x support, will we then have a minimum of `3.3.0` (or whatever release makes sense at the time), rather than `3.0.0`?

**Comment by jrfnl** [comment]
@GaryJones Based on the interesting things which have been happening in PHPCS, my current recommendation would be to have a minimum of `3.3.0`.
* The whole 3.0 range is buggy for external rulesets because of autoload issues.
* 3.1 versions are ok.
* 3.2 adds the awesome new selective ignore syntax, but then forgot to take it into account for all the sniffs, so most of the sniffs (and more so the fixers) are buggy as they don't handle the new annotations well.
* 3.3 is looking good so far.

If nothing else, having `3.3.0` as a minimum would be helpful to our end-users combined with a good upgrade guide. End-users which use a custom ruleset will need to check their ruleset anyway what with the WPCS renamed sniffs being dropped in 2.x, so if we can entice them to update the properties to the new format in one go, it will save them some work.

**Comment by jrfnl** [comment]
PHPCS 3.3.0 will add a new `Generic.PHP.LowerCaseType` sniff which can check that type casts, parameter an return type declarations (non-class based ones) are all lowercase.

While not strictly covered in the Coding Standards handbook, I would suggest adding it to the `core` ruleset, in line with the previously added `Generic.PHP.LowerCaseConstant` and `Generic.PHP.LowerCaseKeyword` sniffs which are already in that ruleset.

**Comment by jrfnl** [comment]
PHPCS 3.3.0 has added a new `PSR12.Classes.ClassInstantiation` sniff which partially covers what the `WordPress.Classes.ClassInstantiation` sniff covers. There were some bugs, but those have now been fixed and will be in the PHPCS 3.3.1 release.

When this ticket becomes active, it is worth checking if some of the check in the WP sniff can be removed in favour of the upstream sniff.

Also see: https://github.com/squizlabs/PHP_CodeSniffer/issues/2053

**Comment by jrfnl** [comment]
Reminder: don't forget to also review and update the `phpcs.xml.dist.sample` file!

**Comment by jrfnl** [comment]
Reminder: drop support for incorrectly passed array properties as per #1368 
Related: https://github.com/squizlabs/PHP_CodeSniffer/pull/1999 / https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/1048#issuecomment-378570641

**Comment by jrfnl** [comment]
Another one which I'd forgotten to add to this thread before:

Anywhere where an array is used containing `T_CLASS, T_ANON_CLASS, T_TRAIT, T_INTERFACE`, it can be replaced with `Tokens::$ooScopeTokens`.

This new array was introduced in PHPCS 3.1.0 via https://github.com/squizlabs/PHP_CodeSniffer/pull/1415

**Comment by GaryJones** [comment]
In terms of timelines, would this dropping of PHPCS 2.x need be done at a major release, or not?

**Comment by jrfnl** [comment]
> In terms of timelines, would this dropping of PHPCS 2.x need be done at a major release, or not?

@GaryJones It depends on your point of view.

Yes:
- Removing significant things should be done in major releases and removing support for PHPCS 2.x could be considered significant in that respect.
- It's a compatibility break which could prevent people from upgrading if they combine using (sniffs from) WPCS with using other external rulesets and the other ruleset(s) is/are not (yet) compatible with PHPCS 3.x.

No:
- PHPCS could be considered as only a dependency of WPCS and something which is an implementation detail.

**Comment by GaryJones** [comment]
I saw a post recently about how Doctrine dropped support for a version of PHP (7.0?) in a non-major release, but that doesn't have the explicit tie-in that we have with PHPCS.

**Comment by GaryJones** [comment]
Just because I looked it up:
 - PHPCS `3.0.0` bumped the minimum level of PHP from `5.1.2` to `5.4.0`.
 - `master` branch (`3.3.0`) of PHPCS is still PHP `5.4.0`.

So WPCS can jump to at least PHP `5.4.0` when we drop PHPCS `2.x`.
